File: app/src/costs/costs.service.ts
import { Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { ICost } from 'src/interfaces/common';
import { Repository } from 'typeorm';

import CostEntity from './costs.entity';
import log, { LogLevels } from 'src/logger';

@Injectable()
export class CostsService {
  constructor(
    @InjectRepository(CostEntity)
    private costsRepository: Repository<CostEntity>,
  ) {}

  async addCost(
    cost: ICost,
    userId: number,
  ): Promise<{
    status: number;
    data: { cost: CostEntity | null; msg: string };
  }> {
    const newCost = new CostEntity();
    newCost.amount = cost.amount;
    newCost.budget_id = cost.budget_id;
    newCost.date = new Date(cost.date);
    newCost.description = cost.description;
    newCost.title = cost.title;
    newCost.year = cost.year;
    newCost.month = cost.month;
    newCost.user_id = userId;
    newCost.group_id = cost.group_id;
    newCost.account_id = cost.account_id;
    newCost.period = cost.year + '-' + String(cost.month).padStart(2, '0');

    try {
      await this.costsRepository.save(newCost);
      return { status: 202, data: { cost: newCost, msg: '' } };
    } catch (err) {
      console.log(err);
      log(`From cost service: ${err}`, LogLevels.ERROR);
      return { status: 500, data: { cost: null, msg: err } };
    }
  }

  async deleteCost(costId: number): Promise<{
    status: number;
    data: { isDeleted: boolean; msg: string };
  }> {
    try {
      await this.costsRepository.delete(costId);
      return { status: 200, data: { isDeleted: true, msg: '' } };
    } catch (err) {
      console.log(err);
      log(`From cost service: ${err}`, LogLevels.ERROR);
      return { status: 500, data: { isDeleted: false, msg: err } };
    }
  }

  // TODO In develop
  async deleteCostByGroupId(group_id: number): Promise<{
    status: number;
    data: { isDeleted: boolean; msg: string; account_id: null | number };
  }> {
    try {
      const costs = await this.costsRepository.findBy({ group_id });
      let account_id = null;
      if (costs && costs.length > 0) {
        account_id = costs[0].account_id;
        await this.costsRepository
          .createQueryBuilder()
          .delete()
          .from('cost')
          .where('group_id=:group_id', { group_id })
          .execute();
      }

      return { status: 200, data: { isDeleted: true, msg: '', account_id } };
    } catch (err) {
      console.log(err);
      log(`From cost service: ${err}`, LogLevels.ERROR);
      return {
        status: 500,
        data: { isDeleted: false, msg: err, account_id: null },
      };
    }
  }

  async getCostsByPeriod(
    period: string,
    userId: number,
    accountID: number,
  ): Promise<{ costs: CostEntity[] | null; graphData: any; msg: string }> {
    try {
      const costs = await this.costsRepository.findBy({
        period,
        user_id: userId,
        account_id: accountID,
      });
      const graphData = this.addGraphData(costs);
      return { costs: this.normalizeCosts(costs), graphData, msg: '' };
    } catch (err) {
      console.log(err);
      log(`From cost service: ${err}`, LogLevels.ERROR);
      return { costs: null, graphData: null, msg: err };
    }
  }

  // TODO Move ti utils
  addGraphData(costs) {
    const graphCosts = [];
    const graphDays = [];
    const days = new Set();
    for (const cost of costs) {
      days.add(new Date(cost.date).getUTCDate());
    }
    for (const day of days) {
      let sum = 0;
      for (const cost of costs) {
        const day2 = new Date(cost.date).getUTCDate();
        if (day === day2) sum += cost.amount;
      }
      graphCosts.push(sum);
      graphDays.push(day);
    }
    graphCosts.reverse();
    graphDays.reverse();
    return {
      days: [1, 2, 3],
      costs: graphCosts,
    };
  }

  normalizeCosts(costs) {
    if (costs.length > 0) {
      const items = [];
      const periods = new Set();
      costs.map((item) => {
        periods.add(item.date.toString());
      });
      let spentByPeriod = 0;
      for (const period of periods) {
        const item = costs.filter((item) => item.date.toString() === period);
        let spentByDay = 0;
        for (const i of item) {
          spentByDay += i.amount;
          spentByPeriod += i.amount;
        }
        items.push({
          period,
          items: item,
          spentByDay,
          spentByThisMonth: spentByPeriod,
        });
      }
      return items.sort((a, b) => {
        if (new Date(a.period) > new Date(b.period)) return -1;
        if (new Date(a.period) < new Date(b.period)) return 1;
        return 0;
      });
    } else {
      return costs;
    }
  }
}

File: analytics/services/costs_service.py
from datetime import datetime

from models.costs_model import Costs
from models.cost_groups_model import CostGroups
from services.db_service import DBService


class CostsService:
    @staticmethod
    def get_costs(user_id: int, period: str, account_id: int) -> dict:
        return CostsService.normalize_costs(DBService.get_costs_by_period(user_id, period, account_id))

    @staticmethod
    def get_costs_sources(user_id: int, account_id: int) -> list:
        return CostsService.normalize_costs_group(DBService.get_costs_groups(user_id, account_id))

    @staticmethod
    def normalize_costs(costs: list) -> dict:
        items = []
        periods = set(cost.date.isoformat() for cost in costs)
        periods = sorted(periods, reverse=True)

        for period in periods:
            spentByDay = 0
            costsInDate = []
            for cost in costs:
                if cost.date.isoformat() == period:
                    costsInDate.append(CostsService.get_public_cost(cost))
                    spentByDay += cost.amount
            items.append({
                'period': period,
                'items': costsInDate,
                'spentByDay': spentByDay,
            })

        graph_data = CostsService.add_graph_data(costs)
        statistics = CostsService.calculate_month_sum(items)
        return {'costs': items, 'graph_data': graph_data, 'statistics': statistics}

    @staticmethod
    def get_public_cost(cost: Costs) -> dict:
        return {
            'amount': cost.amount,
            'id': cost.id,
            'group_id': cost.group_id,
            'account_id': cost.account_id,
            'title': cost.title,
            'year': cost.year,
            'month': cost.month,
            'period': cost.period,
            'description': cost.description,
            'date': cost.date,
            'budget_id': cost.budget_id
        }

    @staticmethod
    def normalize_costs_group(costs_groups: list) -> list:
        return [CostsService.get_public_cost_group(cost_group) for cost_group in costs_groups]

    @staticmethod
    def get_public_cost_group(cost_group: CostGroups) -> dict:
        return {
            'id': cost_group.id,
            'account_id': cost_group.account_id,
            'title': cost_group.title,
            'description': cost_group.description,
            'order': cost_group.order
        }

    @staticmethod
    def add_graph_data(costs: list):
        graph_costs = []
        graph_days = []
        days = set([cost.date.isoformat() for cost in costs])
        days = sorted(days, reverse=False)

        for day in days:
            sum_of_costs = 0
            for cost in costs:
                day2 = cost.date.isoformat()
                if day == day2:
                   sum_of_costs += cost.amount
            graph_costs.append(sum_of_costs)
            graph_days.append(day)
        return {
            'days': graph_days,
            'items': graph_costs,
        }

    @staticmethod
    def calculate_month_sum(costs):
        if not costs:
            return {'spentByMonth': 0}
        sum_of_month = 0
        for cost in costs:
            sum_of_month += cost['spentByDay']

        return {'spentByMonth': sum_of_month}
